ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Storage/LedgerDB/V2/LedgerSeq.hs

```diff
@@ -81,22 +81,44 @@ import Prelude hiding (read)
   LedgerTablesHandles
 -------------------------------------------------------------------------------}
 
+-- | The interface fulfilled by handles on both the InMemory and LSM handles.
 data LedgerTablesHandle m l = LedgerTablesHandle
   { close :: !(m ())
   , duplicate :: !(m (LedgerTablesHandle m l))
-  -- ^ It is expected that this operation takes constant time.
-  , read :: !(LedgerTables l KeysMK -> m (LedgerTables l ValuesMK))
-  , readRange :: !((Maybe (TxIn l), Int) -> m (LedgerTables l ValuesMK))
-  , readAll :: !(m (LedgerTables l ValuesMK))
+  -- ^ Create a copy of the handle.
+  --
+  -- When applying diffs to a table, we will first duplicate the handle, then
+  -- apply the diffs in the copy. It is expected that this operation takes
+  -- constant time.
+  , read :: !(l EmptyMK -> LedgerTables l KeysMK -> m (LedgerTables l ValuesMK))
+  -- ^ Read values for the given keys from the tables, and deserialize them as
+  -- if they were from the same era as the given ledger state.
+  , readRange :: !(l EmptyMK -> (Maybe (TxIn l), Int) -> m (LedgerTables l ValuesMK, Maybe (TxIn l)))
+  -- ^ Read the requested number of values, possibly starting from the given
+  -- key, from the tables, and deserialize them as if they were from the same
+  -- era as the given ledger state.
+  --
+  -- The returned value contains both the read values as well as the last key
+  -- retrieved. This is necessary because the LSM backend uses an alternative
```

Comment:
Some questions that might be worth answering in this comment:

- Is the last key retrieved part of the returned map? 
- What is the user expected to do with the returned "last key retrieved"?
- When does this function returns `Nothing` ?